tests/ts_updown_util.erl

```diff
@@ -0,0 +1,372 @@
+    Failures =
+        lists:foldl(
+          fun(Scenario, FF) ->
+                  run_scenario(Config, Scenario) ++ FF
+          end,
+          [], Scenarios),
+
+    Failures.
+
+
+-spec run_scenario(config(), #scenario{})
+                  -> [#failure_report{}].
+run_scenario(Config,
+             #scenario{table_node_vsn = TableNodeVsn,
+                       query_node_vsn = QueryNodeVsn,
+                       need_table_node_transition = NeedTableNodeTransition,
+                       need_query_node_transition = NeedQueryNodeTransition,
+                       need_pre_cluster_mixed = NeedPreClusterMixed,
+                       need_post_cluster_mixed = NeedPostClusterMixed,
+                       %% for these, we may have invariants in Config:
+                       data = Data_,
+                       ddl = DDL_,
+                       select_vs_expected = SelectVsExpected_}) ->
+    NodesAtVersions0 =
+        [{N, rtdev:node_version(rtdev:node_id(N))} || N <- ?CFG(nodes, Config)],
+
+    %% 0. retreive scenario-invariant data from Config
+    [Data, DDLFmt, SelectVsExpected] =
+        [begin
+             case Supplied of
+                 undefined ->
+                     ?CFG(Item, Config);
+                 Defined ->
+                     Defined
+             end
+         end || {Supplied, Item} <- [{Data_, data}, {DDL_, ddl},
+                                     {SelectVsExpected_, select_vs_expected}]],
+
+    %% 1. Generate a unique table name, produce DDL from the template
+    {_Mega, Sec, Milli} = os:timestamp(),
+    %% scenarios are not runnung in parallel; even so, for the blazing
+    %% fast machines:
+    timer:sleep(1),
+    Table = list_to_binary(fmt("updown_test_~b~b", [Sec, Milli])),
+    DDL = fmt(DDLFmt, [Table]),
+    ct:log("Scenario: table/query_node_vsn: ~p/~p\n"
+           "          need_table_node_transition: ~p\n"
+           "          need_query_node_transition: ~p\n"
+           "          need_pre_cluster_mixed: ~p\n"
+           "          need_post_cluster_mixed: ~p\n"
+           "     DDL: ~p\n"
+           " ~b queries\n", [TableNodeVsn, QueryNodeVsn,
+                             NeedTableNodeTransition, NeedQueryNodeTransition,
+                             NeedPreClusterMixed, NeedPostClusterMixed,
+                             DDL, length(SelectVsExpected)]),
+
+    %% 2. Pick two nodes for create table and subsequent selects
+    {TableNode, NodesAtVersions1} =
+        find_or_make_node_at_vsn(NodesAtVersions0, TableNodeVsn, []),
+    {QueryNode, NodesAtVersions2} =
+        find_or_make_node_at_vsn(NodesAtVersions1, QueryNodeVsn, [TableNode]),
+
+    %% 3. Try to ensure cluster is (not) mixed as hinted but keep the
+    %%    interesting nodes at their versions as set in step 1.
+    NodesAtVersions3 =
+        ensure_cluster(NodesAtVersions2, NeedPreClusterMixed, [TableNode, QueryNode]),
+
+    %% 4. Create table, put data (this step is always assumed to
```

Comment:
This precondition isn't true though - the ASC/DESC tests for 1.4 MUST fail on CREATE TABLE...